benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,281 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <gflags/gflags.h>
+
+
+#ifdef HAS_CUDA
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include "cuda/base/device_guard.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+#ifdef HAS_HIP
+
+
+#include <hip/hip_runtime.h>
+
+
+#include "hip/base/device_guard.hip.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+DEFINE_bool(gpu_timer, false,
+            "use gpu timer based on event. It is valid only when "
+            "executor is cuda or hip");
+
+
+class Timer {
+public:
+    void tic()
+    {
+        assert(tic_called_ == false);
```

Comment:
Not sure if this is useful, as if someone puts a `#define NDEBUG` at some point, it gets optimized out when in `Release`. Maybe you can define your own simple `GKO_ASSERT` macro ?

Reply:
I guess the point would be to avoid this check in release, so as to disturb the timing as little as possible? In case someone suspects issues, they can run in debug and check. Admittedly, the extra assert would not usually matter for the timing, but still, you never know how it might be used later.